### ISSUE TITLE
Non-Debian tools to be pre-installed in ddev-webserver

### ISSUE THREAD

**Issue by rfay** [drud/ddev#2380]
**Is your feature request related to a problem? Please describe.**

Lots of web developers these days need some specific tools inside the ddev-webserver image that aren't available as regular Debian packages:

* gulp
* grunt
* nvm

We should install nvm with a default nodejs instead of installing nodejs separately.

**Comment by chellman** [comment]
I might not the right person to comment because the tradeoffs of running such builds in the container don't work for me; I always run my builds on the host.  That being said, here are a couple thoughts.

I'm not sure I'd want any of these node-based tools installed by default. For projects that initiated with ddev from the beginning, having default tools is fine because you're guaranteeing that everyone's using the same version of node, etc. For my part, I'd rather have good documentation for those that need it on how to install them (I see there's (this)[https://ddev.readthedocs.io/en/stable/users/extend/customizing-images/#adding-extra-dockerfiles-for-webimage-and-dbimage], which could be expanded/separated into a separate page for tools like these).  

Basically I'd want to avoid the performance tradeoff of running in the container (or the double-the-storage issue with mutagen), and mismatches in the installed version of gulp/grunt/whatever. I've experienced issues with Sass builds (e.g.) based on different versions of grunt and libsass succeeding, but having cosmetic differences that mean the files' diffs don't match.  Would definitely want to avoid that, so I'd like the project to determine what's installed, instead of having default versions.

**Comment by mariohernandez** [comment]
For personal use, I agree with @chellman, however, I am a trainer, and having a uniformed environment for all students during a workshop is beneficial.  For this reason, some of the tools I can think of on having installed and configured include:

- NVM
- Latest LTS Node w/NPM
- Gulp

Alternatives to some of the tools above may include:
- Grunt
- Yarn
- Webpack

Many of other dependencies can be installed with NPM if they exist in a project's **composer.json**.

**Comment by rfay** [comment]
Looks like it's going to be time to add composer 2 before long. It's currently in alpha 2.

**Comment by brianyoungblood** [comment]
My insight is a little shallow at the moment even though I've been using dockers for over a year with another wrapper solution. I've found my expectation is composer, sass, compass, and ruby among others mentioned would be baked in but I don't know what the size considerations would be for those that don't need them.

With that, having the documentation include those examples does meet my needs at the moment. I also think people new to ddev and docker find those examples very helpful and allows for users of ddev to learn how to customize those containers later too. I'll give it more thought and swing back for more thoughts if this discussion continues.

**Comment by kszymukowicz** [comment]
I would go only for nvm as I always install gulp or grunt locally anyway.

**Comment by begrafx** [comment]
Perhaps this should all be part of the `ddev config` process. Pick what tools are needed, leave out what isn't?

**Comment by rfay** [comment]
@begrafx Every Debian package is already available to be installed using `webimage_extra_packages` in config.yaml, or with an argument. The problem here is things that aren't packaged as debian packages, like nvm, gulp, grunt, maybe compass.

**Comment by kszymukowicz** [comment]
Because of some edge cases when using nvm (nvm is bash alias setup) I'd prefer to just put in `.ddev/config.yaml` what node version I want to have in web container. The same as with `mysql_version` / `php_version`.

All other things like gulp / grunt / sass / compass can be part of packages.json reqs and installed by node in `node_modules/.bin/`. I am not sure why they should be part of web image?

**Comment by rfay** [comment]
The basic problem is trying to keep the web container from growing without bound. I'll think about it, but don't see an easy way for multiple node versions to be pre-installed. There might be a way to do something like we do with webimage_extra_packages, adding specific recipes that got added based on config.yaml. 

It seems like lots of people want gulp/grunt/sass/compass to be globally installed? I'm not sure myself of best practice, but think your question is quite important.

**Comment by chellman** [comment]
The best practice, or at least a good one in my experience, is to have your node-based tools locked to the project, which itself has its node version locked via a .nvmrc or .node-version file, so that nvm (or [fnm](https://github.com/Schniz/fnm), which I like) can use it.  So I would discourage a global install of all tools based on node.  If you install a project that was built with a different version, you can end up with crazy diffs trying to commit code back into a project otherwise.

So having a node version manager available in the container, for those that want that, would be as far as I would recommend.

**Comment by kszymukowicz** [comment]
@chellman 
_"The best practice, or at least a good one in my experience, is to have your node-based tools locked to the project, which itself has its node version locked via a .nvmrc or .node-version"_

I agree here that this is the best practice in situation when you MUST have multiple node versions in one environment - because you do not use docker or you do not use docker for node. If we go the same way for PHP we can also make `.phprc` and put there a php version which should be used for cli. This is relatively easy to achieve with https://direnv.net/ and I was doing like that before switching to ddev. The node and php version for cli was automatically changed when I opened the project folder. But we do not have such `.phprc` setting in ddev and the right version of php is used.
 
Here we have docker and changing to different node version is relatively easy. 
I am not sure if this is the best practice but for now I use this:

```
RUN apt-get remove -y nodejs
RUN curl -sL https://deb.nodesource.com/setup_10.x | bash -
RUN apt-get install -y nodejs
```

IMO because we can precisely craft the env in ddev I do not think we need to have nvm here. All we need is the right version of node and install all other packages like: gulp / grunt / webpack from package.json level locally.

**Comment by kszymukowicz** [comment]
And I am starting to think that we do not even need node (and anything like nvm, gulp, grunt) if installing node is 3 liner and its pretty fast like 20-30 sec for first install and instant when changed later.
 
Sure I would like to see node version as option in `ddev/config.yaml` because node install is the only reason for me now to have `.ddev/web-build/Dockerfile` but if I need to install it myself its not a big deal also.

**Comment by rfay** [comment]
drush launcher has been updated again, to 0.8.0

**Comment by rfay** [comment]
gulp-cli is in v1.16; other things will wait...

**Comment by rfay** [comment]
Regardless of anything else, nodejs will be v16 for ddev v1.19, see https://github.com/drud/ddev/issues/3375

However, this issue is too long-in-the tooth to be actionable anymore. If you have preferences somewhere in here that you think should be built-in to the ddev-webserver, please open specific issues for them. Closing, but happy to continue the conversation here or elsewhere.
